jd4/cgroup.py

```diff
@@ -10,58 +10,63 @@
 from jd4.log import logger
 from jd4.util import read_text_file, write_text_file
 
-CPUACCT_CGROUP_ROOT = '/sys/fs/cgroup/cpuacct/jd4'
-MEMORY_CGROUP_ROOT = '/sys/fs/cgroup/memory/jd4'
-PIDS_CGROUP_ROOT = '/sys/fs/cgroup/pids/jd4'
+CGROUP2_ROOT = '/sys/fs/cgroup/jd4'
+CGROUP2_DAEMON_ROOT = path.join(CGROUP2_ROOT, 'daemon')
 WAIT_JITTER_NS = 5000000
 
 def try_init_cgroup():
     euid = geteuid()
-    cgroups_to_init = list()
-    if not (path.isdir(CPUACCT_CGROUP_ROOT) and access(CPUACCT_CGROUP_ROOT, W_OK)):
-        cgroups_to_init.append(CPUACCT_CGROUP_ROOT)
-    if not (path.isdir(MEMORY_CGROUP_ROOT) and access(MEMORY_CGROUP_ROOT, W_OK)):
-        cgroups_to_init.append(MEMORY_CGROUP_ROOT)
-    if not (path.isdir(PIDS_CGROUP_ROOT) and access(PIDS_CGROUP_ROOT, W_OK)):
-        cgroups_to_init.append(PIDS_CGROUP_ROOT)
-    if cgroups_to_init:
+    if not (path.isdir(CGROUP2_ROOT) and access(CGROUP2_ROOT, W_OK)):
+        cgroup2_subtree_control = path.join(CGROUP2_ROOT, 'cgroup.subtree_control')
         if euid == 0:
-            logger.info('Initializing cgroup: %s', ', '.join(cgroups_to_init))
-            for cgroup_to_init in cgroups_to_init:
-                makedirs(cgroup_to_init, exist_ok=True)
+            logger.info('Initializing cgroup: %s', CGROUP2_ROOT)
+            write_text_file('/sys/fs/cgroup/cgroup.subtree_control', '+cpu +memory +pids')
+            makedirs(CGROUP2_ROOT, exist_ok=True)
+            write_text_file(cgroup2_subtree_control, '+cpu +memory +pids')
+            makedirs(CGROUP2_DAEMON_ROOT, exist_ok=True)
         elif __stdin__.isatty():
-            logger.info('Initializing cgroup: %s', ', '.join(cgroups_to_init))
-            call(['sudo', 'sh', '-c', 'mkdir -p "{1}" && chown -R "{0}" "{1}"'.format(
-                euid, '" "'.join(cgroups_to_init))])
+            logger.info('Initializing cgroup: %s', CGROUP2_ROOT)
+            call(['sudo', 'sh', '-c',
+                  '''echo "+cpu +memory +pids" > "/sys/fs/cgroup/cgroup.subtree_control" &&
+                     mkdir -p "{1}" &&
+                     chown -R "{0}" "{1}" &&
+                     echo "+cpu +memory +pids" > "{2}" &&
+                     mkdir -p "{3}" &&
+                     chown -R "{0}" "{3}"'''.format(
+                euid, CGROUP2_ROOT, cgroup2_subtree_control, CGROUP2_DAEMON_ROOT)])
         else:
             logger.error('Cgroup not initialized')
 
+    # Put myself into the cgroup that I can write.
```

Comment:
Why is this not needed previously but needed now?

Reply:
On systems with cgroup2 enabled, at the beginning, it is very likely that the process is in a cgroup that it does not have permission to write so it cannot move its children from this initial cgroup to the cgroup used for judging (because leaving a cgroup requires write permissions).